Certificate renewal bot

timothyb89 opened the pull request:
This adds a new `tbot` tool to continuously renew a set of certificates after registering with a Teleport cluster using a similar process to standard node joining.

This makes some modifications to user certificate generation to allow for certificates that can be renewed beyond their original TTL, and exposes new gRPC endpoints:
 * `CreateBot` creates a bot user, role, and a join token
 * `DeleteBot` removes an existing bot user and its resources
 * `GetBotUsers` fetches all Users with bot fields set
 * `GenerateInitialRenewableUserCerts` exchanges a token for a set of certificates with a new `renewable` flag set

A new `tctl` command, `tctl bots add`, creates a bot user and calls `CreateBotJoinToken` to issue a token. A bot instance can then be started using a provided command.

Refer to the RFD for more details: https://github.com/gravitational/teleport/pull/7986/

### Security Considerations

* Renewable certificates: certificates that can be renewed indefinitely (until the backing User is locked or deleted)
  * Several mechanisms are added to limit the security blast radius of these certs:
    * Role impersonation allows split use privilege deescalation: #9561
    * Certificate generation counter validation helps limit utility of stolen renewable certificates: #10098 
* Bot management
  * `CreateBot` and `DeleteBot` gRPC calls create roles which allow role impersonation (i.e. they allow arbitrary privilege escalation). The endpoints do check that the calling user has relevant RBAC permissions to create/delete Roles and Users. We assume that the "create Role" RBAC privilege is effectively root access in Teleport today.

### Outstanding TODOs

* ~~Thorough unit testing~~ Mostly done in #10098
* Implement `tbot init` to configure file ACLs (client side; separate WIP at https://github.com/gravitational/teleport/pull/10289)
* RFD compliance pass
* ~~Validate / resolve outstanding TODOs related to bot locking (don't expect much exciting content here - bots are just users after all)~~ Done in #10098
* Handle CA rotations (client side; will follow up in a separate PR)

timothyb89 committed:
Add certificate renewal bot

This adds a new `tbot` tool to continuously renew a set of
certificates after registering with a Teleport cluster using a
similar process to standard node joining.

This makes some modifications to user certificate generation to allow
for certificates that can be renewed beyond their original TTL, and
exposes new gRPC endpoints:
 * `CreateBotJoinToken` creates a join token for a bot user
 * `GenerateInitialRenewableUserCerts` exchanges a token for a set of
   certificates with a new `renewable` flag set

A new `tctl` command, `tctl bots add`, creates a bot user and calls
`CreateBotJoinToken` to issue a token. A bot instance can then be
started using a provided command.

timothyb89 committed:
Cert bot refactoring pass

* Use role requests to split renewable certs from end-user certs
* Add bot configuration file
* Use `teleport.dev/bot` label
* Remove `impersonator` flag on initial bot certs
* Remove unnecessary `renew` package
* Misc other cleanup

timothyb89 committed:
Do not pass through `renewable` flag when role requests are set

This adds additional restrictions on when a certificate's `renewable`
flag is carried over to a new certificate. In particular, it now also
denies the flag when either role requests are present, or the
`disallowReissue` flag has been previously set.

In practice `disallow-reissue` would have prevented any undesired
behavior but this improves consistency and resolves a TODO.

timothyb89 committed:
Various tbot UX improvements; render SSH config

* Fully flesh out config template rendering
* Fix rendering for SSH configuration templates
* Added `String()` impls for destination types
* Improve certificate renewal logging; show more detail
* Properly fall back to default (all) roles
* Add mode hints for files
* Add/update copyright headers

timothyb89 committed:
Add stubs for tbot init and watch commands

timothyb89 committed:
Add gRPC endpoints for managing bots

* Add `CreateBot`, `DeleteBot`, and `GetBotUsers` gRPC endpoints
* Replace `tctl bot (add|rm|ls)` implementations with gRPC calls
* Define a few new constants, `DefaultBotJoinTTL`, `BotLabel`,
  `BotGenerationLabel`

timothyb89 committed:
Fix outdated destination flag in example tbot command

timothyb89 committed:
Bugfix pass for demo

* Fixed a few nil pointer derefs when using config from CLI args
* Properly create destination if `--destination-dir` flag is used
* Remove improper default on CLI flag
* `DestinationConfig` is now a list of pointers

timothyb89 commented:
I'm trying to avoid adding more onto this PR directly since it's already gigantic, so I'll keep a list of addon PRs here:
* https://github.com/gravitational/teleport/pull/10098

nklaassen commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,607 @@
+
+		// TODO: consider `memory` dest type for testing / ephemeral use / etc
+
+		log.Debug("attempting first connection using initial auth client")
+		authClient, err = authenticatedUserClientFromIdentity(ident, botConfig.AuthServer)
```
this calls `auth.NewClient` which dials in the background, you may not see an error until the first API call. Could try Ping or something if you want to make sure the identity is good

timothyb89 replied:
Oh, good catch! I'll add a ping test.

nklaassen commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,607 @@
+		// TODO: warn if duration < certTTL? would indicate TTL > server's max renewable cert TTL
+		// TODO: error if duration < renewalInterval? next renewal attempt will fail
+
+		// Immediately attempt to reconnect using the new identity (still
+		// haven't persisted the known-good certs).
+		newClient, err := authenticatedUserClientFromIdentity(newIdentity, cfg.AuthServer)
```
this calls `auth.NewClient` which dials in the background, you may not see an error until the first API call. Could try Ping or something if you want to make sure the identity is good

nklaassen commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,607 @@
+func watchCARotations(watcher types.Watcher) {
+	for {
+		select {
+		case event := <-watcher.Events():
+			log.Debugf("CA event: %+v", event)
```
will trigger cert renewal or something here, right?

timothyb89 replied:
Yep, missing a TODO here

nklaassen commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,607 @@
+		return nil, trace.Wrap(err)
+	}
+
+	// The root CA included with the returned user certs will only contain the
+	// Teleport User CA. We'll also need the host CA for future API calls.
+	localCA, err := client.GetClusterCACert()
```
for my own understanding, which api calls are made with the impersonated identity?

timothyb89 replied:
Ideally nothing, unless some end user wants to use the impersonated certs to access Teleport APIs. The "future API calls" referenced in the comment is... copied and pasted from `renewIdentityViaAuth()` so most likely not relevant anymore. These CAs would probably be needed if an end-user wants to access Teleport APIs, though.

nklaassen reviewed (commented):
did a first pass on this

timothyb89 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1826,6 +1927,54 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+
+// CreateBotJoinToken creates a new join token for a certificate renewal bot.
+func (a *ServerWithRoles) CreateBotJoinToken(ctx context.Context, req CreateUserTokenRequest) (types.UserToken, error) {
+	if err := a.action(apidefaults.Namespace, types.KindUser, types.VerbUpdate); err != nil {
```
Oh, interesting. I'll hold off on changing anything here until it finalizes but that feels much better.

zmb3 commented on constants.go:
```diff
@@ -252,6 +255,9 @@ const (
 	ComponentWindowsDesktop = "windows_desktop"
 
+	// ComponentBot is a certificate renewal bot.
+	ComponentBot = "bot"
```
What's the difference between `bot` and `tbot`? Do we need them both?

timothyb89 replied:
D'oh, nope!

I definitely didn't just forget that we already had a constant defined...

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1515,9 +1515,49 @@ func (a *ServerWithRoles) NewKeepAliver(ctx context.Context) (types.KeepAliver,
+	}
+
+	if err := a.authServer.verifyUserToken(token, UserTokenTypeBot); err != nil {
+		return nil, trace.Wrap(err)
```
Should this be `trace.AccessDenied`?

timothyb89 replied:
`verifyUserToken` returns `trace.AccessDenied()` for the relevant denials, and I thought `trace.Wrap()` bubbled up the specific error? Almost all our other calls to `verifyUserToken` behave the same here, so I think it's okay?

zmb3 commented on tool/tbot/auth.go:
```diff
@@ -0,0 +1,197 @@
+	Clock clockwork.Clock
+}
+
+// insecureRegisterClient attempts to connects to the Auth Server using the
```
```suggestion
// insecureAuthClient attempts to connect to the Auth Server using the
```

zmb3 commented on tool/tbot/config/config.go:
```diff
@@ -0,0 +1,254 @@
+)
+
+const (
+	CONFIG_DEFAULT_CERTIFICATE_TTL = 60 * time.Minute
```
```suggestion
	DefaultCertificateTTL = 60 * time.Minute
```

`SCREAMING_CASE` is not typical in Go.

zmb3 commented on tool/tbot/config/config.go:
```diff
@@ -0,0 +1,254 @@
+		}
+	}
+
+	if conf.AuthServer == "" {
```
nit: but I generally prefer to do these checks first, so that we avoid mutations via any of the other `CheckAndSetDefaults` calls if we are going to return an error anyway.

timothyb89 replied:
Good thought, and a quick fix!

zmb3 commented on tool/tbot/config/config_destination.go:
```diff
@@ -0,0 +1,81 @@
+type ConfigKind string
+
+const (
+	CONFIG_KIND_SSH ConfigKind = "ssh"
```
```suggestion
	KindSSH ConfigKind = "ssh"
```

We're already in package `config`, so this should be sufficient.

zmb3 commented on tool/tbot/config/config_destination.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+// AllKinds lists all valid config kinds, intended for validation purposes.
+var AllKinds = [...]ConfigKind{CONFIG_KIND_SSH, CONFIG_KIND_TLS}
```
Does this need to be exported?

If you use it to validate in other packages, I would consider unexporting this, and creating an exported validation func that reads but does not modify `allKinds`.

It's unlikely to happen, but leaving this exported allows other package to mutate it, which could break things.

timothyb89 replied:
It's not needed outside the package yet, so I'll just leave it unexported for now, we can always add a validation function like you suggested if needed later.

zmb3 commented on tool/tbot/config/config_storage.go:
```diff
@@ -0,0 +1,44 @@
+
+import "github.com/gravitational/trace"
+
+const STORAGE_DEFAULT_PATH = "/var/lib/teleport/bot"
```
How about something like this?

```suggestion
const DefaultStoragePath = filepath.Join(defaults.DataDir, "bot")
```

You'll probaby have to drop the const.

timothyb89 replied:
I like that a lot, and I'll unexport it to avoid any mutation (can add an accessor function if needed later)

zmb3 commented on tool/tbot/config/configtemplate_ssh.go:
```diff
@@ -0,0 +1,215 @@
+	"golang.org/x/crypto/ssh"
+)
+
+const DEFAULT_PROXY_PORT = 3023
```
Do we not already have a constant for this?

timothyb89 replied:
Ah, yup, `defaults.SSHProxyListenPort`

zmb3 commented on tool/tbot/utils/destination.go:
```diff
@@ -0,0 +1,38 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package utils
```
- can you think of a better name than `utils`? It's a slippery slope and ends up becoming a bag of misc code.
- Add a newline between the license and the package clause so that `go doc` doesn't interpret the license as documentation.

timothyb89 replied:
Fair point, I'll just rename it to `destination` - just needs to be its own package to prevent import cycles.

Those load-bearing newlines are fun!

zmb3 commented on tool/tctl/common/bots_command.go:
```diff
@@ -0,0 +1,249 @@
+	"github.com/gravitational/trace"
+)
+
+const BOT_LABEL = "teleport.dev/bot"
```
Worth putting this in api/types/constants.go with `OriginLabel`?

timothyb89 replied:
Looks like I already made a constant for it (`constants.BotLabel`) and maybe even more importantly it's not needed in tctl at all anymore, so I'll just remove it.

zmb3 reviewed (commented):
First pass.

timothyb89 commented on api/client/proto/authservice.proto:
```diff
@@ -358,6 +358,62 @@ message CreateResetPasswordTokenRequest {
+    // RoleName is the name of the associated bot role.
+    string RoleName = 2 [ (gogoproto.jsontag) = "role_name" ];
+    // Token is the name of the join token for the bot.
+    string Token = 3 [ (gogoproto.jsontag) = "token" ];
```
I can certainly change it, but I believe this is consistent with our other APIs since we almost exclusively refer to tokens by their name.

timothyb89 commented on api/client/proto/authservice.proto:
```diff
@@ -358,6 +358,62 @@ message CreateResetPasswordTokenRequest {
+// certificates from a bot join token.
+message RenewableCertsRequest {
+    // Token is a bot join token.
+    string Token = 1 [ (gogoproto.jsontag) = "token" ];
```
Nope, for IAM joining you still pass along a token name, but rather than a random/secure value it's defined in advance.

timothyb89 replied:
Should probably clarify this more - we'll be outright removing `RenewableCertsRequest` and `GenerateInitialRenewableUserCerts` with @nklaassen's work on IAM joining since it'll join using the usual `auth.Register()` flow.

timothyb89 commented on api/client/proto/authservice.proto:
```diff
@@ -358,6 +358,62 @@ message CreateResetPasswordTokenRequest {
 
+// CreateBotJoinTokenRequest is a request to create a reset password token.
+message CreateBotJoinTokenRequest {
+    // Name is the bot's user name.
```
Hrm, good question. That is confusing but it should be the user name.

I kind of want to just remove `CreateBotJoinToken` now that we have `CreateBot` and `DeleteBot` gRPC calls.

timothyb89 commented on api/client/proto/authservice.proto:
```diff
@@ -358,6 +358,62 @@ message CreateResetPasswordTokenRequest {
+    repeated string Roles = 4 [ (gogoproto.jsontag) = "roles" ];
+}
+
+// CreateBotResponse returns required details for
```
Whoops, will fix!

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,252 @@
+	}
+
+	// TODO: deleteUserTokens?
+	_, err = s.Identity.CreateUserToken(ctx, token)
```
Uhh, no good reason, will fix!

timothyb89 commented on lib/auth/usertoken.go:
```diff
@@ -382,6 +403,7 @@ func formatUserTokenURL(proxyHost string, tokenID string, reqType string) (strin
 		u.Path = fmt.Sprintf("/web/recovery/steps/%v/verify", tokenID)
 	}
 
+	// TODO: case for UserTokenTypeBot?
```
That's probably the best way forward if we continue using user tokens. That said, if we switch to regular provisioning tokens  it might be a moot point.

timothyb89 committed:
Address first wave of review feedback

Fixes the majority of smaller issues caught by reviewers, thanks all!

timothyb89 committed:
Add doc comments for bot.go functions

timothyb89 committed:
Return the token TTL from CreateBot

timothyb89 commented on lib/auth/auth.go:
```diff
@@ -997,6 +1004,7 @@ func (a *Server) generateUserCert(req certRequest) (*proto.Certs, error) {
 		AWSRoleARNs:     roleARNs,
 		ActiveRequests:  req.activeRequests.AccessRequests,
 		DisallowReissue: req.disallowReissue,
+		Renewable:       req.renewable,
```
It needs to be set in the `certRequest{...}` which can't be manipulated by users directly, if that's what you mean. `generateUserCerts()` is the only function that writes to the `renewable` flag (via a `certRequestRenewable()` cert request function which itself is only ever set via `GenerateInitialRenewableUserCerts()`).

timothyb89 replied:
Small change now, the `certRequestRenewable()` function has been removed, now only `generateInitialRenewableUserCerts()` sets the flag on new certs and does so explicitly. (`generateUserCerts()` still carries over the flag on renewed certs in certain cases)

nklaassen replied:
thanks, I thought it was coming from the client over gRPC for some reason

timothyb89 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1565,12 +1605,39 @@ func (a *ServerWithRoles) determineDesiredRolesAndTraits(req proto.UserCertsRequ
+}
+
+func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserCertsRequest, allowNopUser bool, opts ...certRequestOption) (*proto.Certs, error) {
 	var err error
 
 	// this prevents clients who have no chance at getting a cert and impersonating anyone
 	// from enumerating local users and hitting database
-	if !a.hasBuiltinRole(string(types.RoleAdmin)) && !a.context.Checker.CanImpersonateSomeone() && req.Username != a.context.User.GetName() {
+	if !a.hasBuiltinRole(string(types.RoleAdmin)) &&
+		!a.context.Checker.CanImpersonateSomeone() &&
+		req.Username != a.context.User.GetName() &&
+		!(allowNopUser && isNopUser(a.context.User)) {
 		return nil, trace.AccessDenied("access denied: impersonation is not allowed")
 	}
```
Yeah, I'd previously been worried that the bot onboarding case had too much overlap with regular cert issuing (and maybe that was more true before role requests and the split between renewable certs and role request'd certs).

Anyway, now the renewable certs are trivial and if we trim `generateUserCerts()` down to exactly what we need while issuing the initial cert, it's just a bit of dressing around a call to `authServer.generateUserCert()`. I've put that logic in a new `generateInitialRenewableUserCerts()` and completely removed all the Nop user stuff.

(also ended up removing `certRequestRenewable()` entirely)

timothyb89 committed:
Split initial user cert issuance from `generateUserCerts()`

Issuing initial renewable certificate ended up requiring a lot of
hacks to skip checks that prevented anonymous bots from getting
certs even though we'd verified their identity elsewhere (via token).

This reverts all those hacks and splits initial bot cert logic into a
dedicated `generateInitialRenewableUserCerts()` function which should
make the whole process much easier to follow.

timothyb89 committed:
Set bot traits to silence log messages

timothyb89 committed:
tbot log message consistency pass

timothyb89 committed:
Resolve lints

timothyb89 committed:
Add config tests

timothyb89 committed:
Remove CreateBotJoinToken endpoint

Users should instead use the CreateBot/DeleteBot endpoints.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 commented on lib/auth/grpcserver.go:
```diff
@@ -2092,7 +2196,7 @@ func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req prot
 
 	// Generate the cert.
-	certs, err := actx.generateUserCerts(ctx, req, certRequestMFAVerified(mfaDev.Id), certRequestClientIP(clientIP))
+	certs, err := actx.generateUserCerts(ctx, req, false, certRequestMFAVerified(mfaDev.Id), certRequestClientIP(clientIP))
```
It's now been removed with the first-time renewable cert onboarding process spun out into its own function.

timothyb89 committed:
Create a fresh private key for every impersonated identity renewal

timothyb89 committed:
Hide `config` subcommand

timothyb89 committed:
Rename bot label prefix to `teleport.internal/`

nklaassen commented on tool/tbot/config/configtemplate.go:
```diff
@@ -0,0 +1,118 @@
+		return nil
+	}
+
+	type rawTemplate TemplateConfig
+	if err := node.Decode((*rawTemplate)(c)); err != nil {
+		return err
+	}
+
+	return nil
```
I don't understand the typecast, could probably use a comment.

Also the `if` is doing nothing and fails lint, you can just `return trace.Wrap(err)`

timothyb89 replied:
Ah, I just didn't copy and paste the comment everywhere. It's derived from https://abhinavg.net/posts/flexible-yaml/#make-it-flexible to allow optional parameters for templates

nklaassen commented on tool/tbot/config/configtemplate_ssh.go:
```diff
@@ -0,0 +1,215 @@
+		return trace.Wrap(err)
+	}
+
+	proxies, err := authClient.GetProxies()
```
You may be able to get this info easier from `authClient.Ping()`. In theory I think all proxies should have the same public addr

timothyb89 replied:
Nice, that's a lot simpler, thanks!

nklaassen commented on tool/tbot/config/destination_directory.go:
```diff
@@ -0,0 +1,78 @@
+	// override (we want to use standard unmarshal behavior for the full
+	// struct)
+	type rawDirectory DestinationDirectory
+	if err := node.Decode((*rawDirectory)(dd)); err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return trace.Wrap(node.Decode((*rawDirectory)(dd)))
```

nklaassen commented on tool/tbot/config/destination_directory.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+func (dd *DestinationDirectory) Write(name string, data []byte, modeHint destination.ModeHint) error {
+	// TODO: honor modeHint?
```
for now just return an error if it's not what you expect

timothyb89 replied:
I think I'm just going to remove `modeHint` in the `tbot init` addon PR, it's ended up unused and I don't see any reason for us to write non-secret values.

nklaassen commented on tool/tbot/config/destination_memory.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	type rawMemory DestinationMemory
+	if err := node.Decode((*rawMemory)(dm)); err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return trace.Wrap(node.Decode((*rawMemory)(dm)))
```

nklaassen commented on tool/tbot/identity/identity.go:
```diff
@@ -0,0 +1,388 @@
+	}{
+		{TLSCertKey, id.TLSCertBytes, destination.ModeHintSecret},
+		{SSHCertKey, id.CertBytes, destination.ModeHintSecret},
+		{TLSCACertsKey, bytes.Join(id.TLSCACertsBytes, []byte("$")), destination.ModeHintSecret},
+		{SSHCACertsKey, bytes.Join(id.SSHCACertBytes, []byte("\n")), destination.ModeHintSecret},
```
any rationale for the different separators? I blindly changed both to `"$"` [here](https://github.com/gravitational/teleport/pull/10397) because SSH CA certs seem to have a terminal newline that was messing up the splitter

timothyb89 replied:
Not especially, nope. If that's simpler I say just switch it.

timothyb89 committed:
Use types.NewRole() to create bot roles

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,279 @@
+
+// createBotRole creates a role from a bot template with the given parameters.
+func createBotRole(ctx context.Context, s *Server, botName string, resourceName string, roleRequests []string) error {
+	return s.UpsertRole(ctx, &types.RoleV4{
```
Fixed, looks like Role is missing `SetMetadata()` so setting our labels is a bit ugly, oh well.

nklaassen replied:
you could implement `SetMetadata()`, it's probably just not there because there was no caller

timothyb89 replied:
I've gone ahead and implemented `SetMetadata()`

timothyb89 committed:
Clean up error handling in custom YAML unmarshallers

Also, add notes about the supported YAML shapes.

timothyb89 committed:
Fetch proxy host via gRPC Ping() instead of GetProxies()

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,252 @@
+		Spec: types.RoleSpecV4{
+			Options: types.RoleOptions{
+				// TODO: inherit TTLs from cert length?
+				MaxSessionTTL: types.Duration(12 * time.Hour),
```
I don't have any particular feelings one way or the other on the `MaxSessionTTL` value, so maybe best to just leave it unspecified.

timothyb89 commented:
I think this is about as ready as it's going to get for review, though it's a bit of a monster so apologies in advance to the reviewers.

A rough review/merge plan that hopefully makes sense:
* Merge the backend leaf PRs into this one, which can be merged in any order:
  * https://github.com/gravitational/teleport/pull/10098 (also resolves the failing unit test here)
  * Nic's work: https://github.com/gravitational/teleport/pull/10360 and https://github.com/gravitational/teleport/pull/10397
* Then, merge this PR into master
* Last, work on merging + backporting the client-side changes (`tbot init`, CA rotation support, etc). These are still in progress and will be merged in the coming weeks.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

zmb3 commented on Makefile:
```diff
@@ -208,6 +208,10 @@ $(BUILDDIR)/teleport: ensure-webassets bpf-bytecode rdpclient
 
+.PHONY: $(BUILDDIR)/tbot
+$(BUILDDIR)/tbot:
+	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG)" -o $(BUILDDIR)/tbot $(BUILDFLAGS) ./tool/tbot
```
Do you get different behavior if you omit the PAM/FIPS tags? If not, can we do without them?

timothyb89 replied:
I was about to say "we can just drop them" but I suppose it's possible the FIPS tag is honored in some of the teleport libraries we use. PAM is almost certainly unnecessary but I included it since `tsh` does for some reason as well.

timothyb89 replied:
I've removed the PAM tag. I'm wary of touching the FIPS tag since I'm not sure what it touches. My gut feeling is that if `tsh` cares, we should too.

zmb3 commented on api/client/client.go:
```diff
@@ -689,6 +689,51 @@ func (c *Client) CreateResetPasswordToken(ctx context.Context, req *proto.Create
+
+// GetBotUsers fetches all bot users.
+func (c *Client) GetBotUsers(ctx context.Context) ([]types.User, error) {
+	stream, err := c.grpc.GetBotUsers(context.TODO(), &proto.GetBotUsersRequest{}, c.callOpts...)
```
```suggestion
	stream, err := c.grpc.GetBotUsers(ctx, &proto.GetBotUsersRequest{}, c.callOpts...)
```

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1610,6 +1662,16 @@ service AuthService {
+    // DeleteBot deletes a bot user.
+    rpc DeleteBot(DeleteBotRequest) returns (google.protobuf.Empty);
+    // GetBotUsers gets all users with bot labels.
+    rpc GetBotUsers(GetBotUsersRequest) returns (stream types.UserV2);
```
Curious, why does this leverage streaming?

timothyb89 replied:
`GetUsers()` does and I just cloned its declaration

zmb3 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+	}
+	meta.Labels[types.BotLabel] = botName
+
+	rolev4, ok := role.(*types.RoleV5)
```
```suggestion
	rolev5, ok := role.(*types.RoleV5)
```

zmb3 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+
+	resourceName := botResourceName(req.Name)
+
+	// Ensure existing resources don't already exist.
```
```suggestion
	// Ensure conflicting resources don't already exist.
```

zmb3 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+
+func (s *Server) deleteBot(ctx context.Context, botName string) error {
+	// TODO:
+	// remove any locks for the bot's impersonator role?
```
Do we have a separate issue where we're tracking locking of bots?

timothyb89 replied:
I've mostly resolved bot locking in https://github.com/gravitational/teleport/pull/10098, I think this is the only remaining open question. I'm inclined to not delete any locks automatically since that would be new (and maybe unexpected) behavior since I don't think we delete locks when e.g. removing a user or role today.

If you agree I can just remove the TODO.

timothyb89 replied:
I've replaced the TODO here with a note explaining why locks aren't currently deleted. Let me know if you'd prefer to see a behavior change instead!

zmb3 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	if req.Type != UserTokenTypeBot {
+		return nil, trace.BadParameter("invalid bot token request type")
```
```suggestion
		return nil, trace.BadParameter("invalid bot token request type %q", req.Type)
```
```suggestion
		return nil, trace.BadParameter("invalid bot token request type")
```

timothyb89 replied:
This has been refactored away due to Nic's work.

zmb3 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+		return nil, trace.Wrap(err)
+	}
+
+	// TODO: audit log event. partially implemented, may need events.proto and
```
For things like this maybe you want to remove the TODO and just track this in a separate issue.

timothyb89 replied:
I've created https://github.com/gravitational/teleport/issues/10477 to track missing audit events

zmb3 commented on tool/tbot/config/configtemplate.go:
```diff
@@ -0,0 +1,122 @@
+		switch simpleTemplate {
+		case TemplateSSHClientName:
+			c.SSHClient = &TemplateSSHClient{}
+			fmt.Println("no params, using defaults")
```
Did you mean for this to be here or was this for debugging?

timothyb89 replied:
Yup, apologies - Nic's actually resolved this already in https://github.com/gravitational/teleport/pull/10397

zmb3 commented on tool/tbot/config/configtemplate_ssh.go:
```diff
@@ -0,0 +1,194 @@
+	// TODO: ideally it'd be nice to fetch this dynamically
+	// TODO: eventually we could consider including `tsh proxy`
+	// functionality and sidestep this entirely.
+	proxyPort := fmt.Sprint(c.ProxyPort)
```
I prever `strconv.Itoa` for something as simple as int->string.

r0mant commented on tool/tbot/identity/identity.go:
```diff
@@ -0,0 +1,388 @@
+	// TODO: host certs use CertExtensionAuthority, client certs use CertExtensionTeleportRouteToCluster
+	// (or at least, teleport-route-to-cluster _appears_ to be a sane cluster name, but I have no idea
+	// how it is actually set.)
+	//clusterName := cert.Permissions.Extensions[utils.CertExtensionAuthority]
```
Do not leave commented out code.

r0mant commented on tool/tbot/identity/identity.go:
```diff
@@ -0,0 +1,388 @@
+}
+
+func LoadIdentity(d destination.Destination) (*Identity, error) {
+	// TODO: encode the whole thing using the identityfile package?
```
It does look like there's a lot of functionality overlap tbh. Any way to reuse some of the identityfile-provided functions?

r0mant commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,631 @@
+	authServerEnvVar = "TELEPORT_AUTH_SERVER"
+	tokenEnvVar      = "TELEPORT_BOT_TOKEN"
+
+	nologinPrefix = "-teleport-nologin-"
```
Don't we already have this constant somewhere? It's definitely used, so if there's no command constant, can we make one?

timothyb89 replied:
We don't, but I'll add one.

r0mant commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,631 @@
+	// Start by loading the bot's primary destination.
+	dest, err := botConfig.Storage.GetDestination()
+	if err != nil {
+		return trace.WrapWithMessage(err, "could not read bot storage destination from config")
```
Just curious why not just trace.Wrap(err, "xxx"). Same outcome but a bit shorter.

timothyb89 replied:
Oh, good to know! I didn't realize it supported formatting args. I'll replace it everywhere.

r0mant commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,631 @@
+		}
+
+		// Attach the ssh public key.
+		//ident.SSHPublicKeyBytes = sshPublicKey
```
Remove commented code.

r0mant commented on tool/tbot/main.go:
```diff
@@ -0,0 +1,631 @@
+		}
+
+		log.Infof("Persisted new certificates to disk. Next renewal in approximately %s", cfg.RenewInterval)
+		<-ticker.C
```
Should this select on a context to be able to cancel?

timothyb89 replied:
Yes, that'll be useful for graceful shutdown support as well.

timothyb89 committed:
Update lib/auth/bot.go

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

timothyb89 commented on tool/tbot/identity/identity.go:
```diff
@@ -0,0 +1,388 @@
+		}
+	}
+
+	// check permissions on certificate
+	// TODO: do we care to even verify this at all?
+	// if len(cert.Permissions.Extensions) == 0 {
+	// 	return nil, trace.BadParameter("extensions: misssing needed extensions for host roles")
+	// }
+
+	// roleString := cert.Permissions.Extensions[utils.CertExtensionRole]
+	// if roleString == "" {
+	// 	return nil, trace.BadParameter("misssing cert extension %v", utils.CertExtensionRole)
+	// }
+	// roles, err := types.ParseTeleportRoles(roleString)
+	// if err != nil {
+	// 	return nil, trace.Wrap(err)
+	// }
+	// foundRoles := len(roles)
+	// if foundRoles != 1 {
+	// 	return nil, trace.Errorf("expected one role per certificate. found %d: '%s'",
+	// 		foundRoles, roles.String())
+	// }
```
This is staged for removal in https://github.com/gravitational/teleport/pull/10289 already - do you think it's something we can defer to that PR or is it important to clear it up now?

timothyb89 commented on tool/tbot/config/configtemplate.go:
```diff
@@ -0,0 +1,122 @@
+		switch simpleTemplate {
+		case TemplateSSHClientName:
+			c.SSHClient = &TemplateSSHClient{}
+			fmt.Println("no params, using defaults")
```
yup, apologies - Nic actually caught and resolved this already in https://github.com/gravitational/teleport/pull/10397

timothyb89 commented on tool/tbot/config/config_destination.go:
```diff
@@ -0,0 +1,79 @@
+	"github.com/gravitational/trace"
+)
+
+type Kind string
```
This is resolved as part of the identity refactoring in https://github.com/gravitational/teleport/pull/10289

r0mant replied:
@timothyb89 @nklaassen If some of the comments are addressed by other PRs going to this branch, can we please merge them before merging this one?

timothyb89 replied:
I've now flattened all the PRs into this one (save for https://github.com/gravitational/teleport/pull/10289 which is coming next week)

timothyb89 commented on tool/tbot/auth.go:
```diff
@@ -0,0 +1,198 @@
+
+	// If no CA was found, then create a insecure connection to the Auth Server,
+	// otherwise use the CA on disk to validate the Auth Server.
+	if trace.IsNotFound(err) {
```
I don't disagree, it's a direct copy of `insecureRegisterClient` from the regular node join process and we could've properly adapted it for our usecase here.

In any case, Nic's change entirely deletes this file: https://github.com/gravitational/teleport/pull/10397

timothyb89 commented on lib/tlsca/ca.go:
```diff
@@ -603,6 +618,11 @@ func FromSubject(subject pkix.Name, expires time.Time) (*Identity, error) {
+		case attr.Type.Equal(RenewableCertificateASN1ExtensionOID):
+			val, ok := attr.Value.(string)
+			if ok {
+				id.Renewable, _ = strconv.ParseBool(val)
```
I'll replace this with a direct comparison to `types.True` as we do for the other bool flags

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+	}
+	meta.Labels[types.BotLabel] = botName
+
+	rolev4, ok := role.(*types.RoleV5)
```
Fair point, I quickly changed it to fix a compile error after pulling in the latest master and clearly my brain wasn't fully working. Will change!

timothyb89 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1957,6 +2089,46 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+// CreateBot creates a new certificate renewal bot and returns a join token.
+func (a *ServerWithRoles) CreateBot(ctx context.Context, req *proto.CreateBotRequest) (*proto.CreateBotResponse, error) {
+	// Note: this creates a role with role impersonation privileges for all
+	// roles listed in the request and doesn't attempt to verify that the
+	// current user has permissions for those embedded roles. We assume that
```
Yes, it can. From what I can, tell, though, `VerbCreate` on `Role` is effectively root-level in Teleport today for exactly this reason, so at least I don't think we're introducing a _new_ path to privilege escalation.

We do check for `VerbCreate` on both `User` and `Role` here so I _think_ we have our bases covered but definitely let me know if I've missed something, this is a pretty scary bit of code IMO.

r0mant replied:
@timothyb89 I'm curious if you have considered making a bot a "resource" like any other object in the cluster. Then it could just be managed by usual tctl create/get/rm commands, we could have a `types.KindBot` which would grant users explicit permissions to manage them, and would be easy to reason about "bot" as a "resource" in general. Probably a bit too late to ask design questions at this point but I'm curious if this has ever been considered.

timothyb89 replied:
That was part of Zac's original design, actually, and I ended up removing it.

Ultimately we needed to reuse so much of Teleport's traditional auth infrastructure (necessitating a backing `User` and `Role`) that the entire `Bot` resource existed solely to be created and removed, alongside the other resources it still needed. Removing the resource was a surprisingly large negative diff and didn't end up breaking any functionality because it just wasn't used.

I wouldn't be surprised to see it come back eventually if we eventually add more of bot-specific metadata. This whole "stuffing bot stuff into `User` labels" strategy is probably not great in the long term .

timothyb89 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1646,9 +1646,115 @@ func (a *ServerWithRoles) NewKeepAliver(ctx context.Context) (types.KeepAliver,
+	}
+
+	if err := a.authServer.DeleteUserToken(ctx, req.Token); err != nil {
+		log.Warnf("could not delete user token %v after generating certs: %v",
```
I think the logic here is that the token will expire on its own anyway, but at any rate this is totally refactored in Nic's backend PR: https://github.com/gravitational/teleport/pull/10360

timothyb89 commented on tool/tbot/auth.go:
```diff
@@ -0,0 +1,198 @@
+// pin, a connection will be re-established and the root CA will be used to
+// validate the certificate presented. If both conditions hold true, then we
+// know we are connecting to the expected Auth Server.
+func pinAuthClient(params RegisterParams) (*auth.Client, []*x509.Certificate, error) {
```
That's a fair point, luckily Nic has already deleted this in https://github.com/gravitational/teleport/pull/10397

timothyb89 commented on tool/tbot/identity/identity.go:
```diff
@@ -0,0 +1,388 @@
+	SSHCACertBytes [][]byte
+	// KeySigner is an SSH host certificate signer
+	KeySigner ssh.Signer
+	// Cert is a parsed SSH certificate
+	Cert *ssh.Certificate
+	// XCert is X509 client certificate
+	XCert *x509.Certificate
```
I like that, I've already renamed a few of these fields as part of https://github.com/gravitational/teleport/pull/10289 so if it's alright with you, I'd like to make that change over there.

r0mant replied:
Sounds good.

timothyb89 committed:
Fix some review comments

timothyb89 committed:
Add renewable certificate generation checks (#10098)

* Add renewable certificate generation checks

This adds a new validation check for renewable certificates that
maintains a renewal counter as both a certificate extension and a
user label. This counter is used to ensure only a single certificate
lineage can exist: for example, if a renewable certificate is stolen,
only one copy of the certificate can be renewed as the generation
counter will not match

When renewing a certificate, first the generation counter presented
by the user (via their TLS identity) is compared to a value stored
with the associated user (in a new `teleport.dev/bot-generation`
label field). If they aren't equal, the renewal attempt fails.
Otherwise, the generation counter is incremented by 1, stored to the
database using a `CompareAndSwap()` to ensure atomicity, and set on
the generated certificate for use in future renewals.

* Add unit tests for the generation counter

This adds new unit tests to exercise the generation counter checks.

Additionally, it fixes two other renewable cert tests that were
failing.

* Remove certRequestGeneration() function

* Emit audit event when cert generations don't match

* Fully implement `tctl bots lock`

* Show bot name in `tctl bots ls`

* Lock bots when a cert generation mismatch is found

* Make CompareFailed respones from validateGenerationLabel() more actionable

* Update lib/services/local/users.go

Co-authored-by: Nic Klaassen <nic@goteleport.com>

* Backend changes for tbot IoT and AWS joining (#10360)

* backend changes

* add token permission check

* pass ctx from caller

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

* fix comment typo

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

* use UserMetadata instead of Identity in RenewableCertificateGenerationMismatch event

* Client changes for tbot IoT joining (#10397)

* client changes

* delete replaced APIs

* delete unused tbot/auth.go

* add license header

* don't unecessarily fetch host CA

* log fixes

* s/tunnelling/tunneling/

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* auth server addresses may be proxies

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* comment typo fix

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* move *Server methods out of auth_with_roles.go (#10416)

Co-authored-by: Tim Buckley <tim@goteleport.com>

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>
Co-authored-by: Tim Buckley <tim@goteleport.com>

Co-authored-by: Roman Tkachenko <roman@goteleport.com>
Co-authored-by: Tim Buckley <tim@goteleport.com>
Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

Co-authored-by: Nic Klaassen <nic@goteleport.com>
Co-authored-by: Roman Tkachenko <roman@goteleport.com>
Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

timothyb89 commented on tool/tbot/config/configtemplate_ssh.go:
```diff
@@ -0,0 +1,194 @@
+
+	dest, err := destination.GetDestination()
+	if err != nil {
+		return err
```
That's an embarrassing miss, will fix.

timothyb89 committed:
Address another batch of review feedback

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+	} else {
+		label, ok := user.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete user %s that is missing label %s", resourceName, types.BotLabel)
```
We don't - I've left a more detailed reply at https://github.com/gravitational/teleport/pull/10099#discussion_r810256017 but while users deleting the labels would cause minor confusion, I think both the error output and resolution steps are reasonably clear.

I'll go ahead and add some more resolution advice to these errors to play it safe, though ("delete the (user|role) manually if desired")

r0mant commented on api/types/constants.go:
```diff
@@ -366,3 +366,11 @@ const (
+
+const (
+	// BotLabel is a label used to identify a resource used by a certificate renewal bot.
+	BotLabel = "teleport.internal/bot"
+
+	// BotGenerationLabel is a label used to record the certificate generation counter.
+	BotGenerationLabel = "teleport.internal/bot-generation"
```
Okay, makes sense.

timothyb89 committed:
Addres another batch of review feedback

Add `Role.SetMetadata()`, simplify more `trace.WrapWithMessage()`
calls, clear some TODOs and lints, and address other misc feedback
items.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Fix lint

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,286 @@
+	// 	},
+	// }); err != nil {
+	// 	log.WithError(err).Warn("Failed to emit create reset password token event.")
+	// }
```
I've removed the commented code and filed an issue regarding missing audit log events: https://github.com/gravitational/teleport/issues/10477

timothyb89 committed:
Add missing doc comments to SaveIdentity / LoadIdentity

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Remove pam tag from tbot build

timothyb89 committed:
Update note about bot lock deletion

r0mant commented on lib/auth/authclient/authclient.go:
```diff
@@ -0,0 +1,114 @@
+	TLS *tls.Config
+	// SSH is client SSH config.
+	SSH *ssh.ClientConfig
+	// AuthAddrs is a list of possible auth or proxy server addresses.
```
```suggestion
	// AuthServers is a list of possible auth or proxy server addresses.
```

r0mant commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,466 @@
+}
+
+// createBot creates a new certificate renewal bot from a bot request.
+func (s *Server) createBot(ctx context.Context, req *proto.CreateBotRequest) (*proto.CreateBotResponse, error) {
```
Should this function also validate that all roles specified in the CreateBotRequest are valid and exist in the system? Ignore if it's already validated somewhere and I just missed it.

timothyb89 replied:
It doesn't, I'll add a quick check

r0mant commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,466 @@
+			// If this fails it's likely to be some miscellaneous competing
+			// write. The request should be tried again - if it's malicious,
+			// someone will get a generation mismatch and trigger a lock.
+			return trace.WrapWithMessage(err, "Database comparison failed, try the request again")
```
Would `trace.CompareFailed` be a more appropriate error to return?

r0mant commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,466 @@
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := s.UpsertLock(context.Background(), lock); err != nil {
```
```suggestion
		if err := s.UpsertLock(ctx, lock); err != nil {
```

r0mant commented on lib/auth/bot.go:
```diff
@@ -0,0 +1,466 @@
+		// If this fails it's likely to be some miscellaneous competing
+		// write. The request should be tried again - if it's malicious,
+		// someone will get a generation mismatch and trigger a lock.
+		return trace.WrapWithMessage(err, "Database comparison failed, try the request again")
```
`trace.CompareFailed`

r0mant commented on lib/auth/join.go:
```diff
@@ -93,10 +95,41 @@ func (a *Server) RegisterUsingToken(ctx context.Context, req *types.RegisterUsin
+		default:
+			// delete bot join tokens so they can't be re-used
+			if err := a.DeleteToken(ctx, provisionToken.GetName()); err != nil {
+				log.WithError(err).Warnf("could not delete bot provision token %q after generating certs",
```
Nit:
```suggestion
				log.WithError(err).Warnf("Could not delete bot provision token %q after generating certs.",
```

r0mant commented on lib/auth/join.go:
```diff
@@ -93,10 +95,41 @@ func (a *Server) RegisterUsingToken(ctx context.Context, req *types.RegisterUsin
+		// bots use this endpoint but get a user cert
+		// botResourceName must be set, enforced in CheckAndSetDefaults
+		botResourceName := provisionToken.GetBotName()
+		expires := a.GetClock().Now().Add(defaults.DefaultRenewableCertTTL)
```
Why does this generate the cert with default cert TTL instead of the one the bot is configured with?

timothyb89 replied:
It's somewhat arbitrary. We don't pass along a desired TTL for the initial certs via the token request here which complicates things. That said, in practice bots immediately attempt to renew certs after joining, and do request a user-configurable TTL then. So for however long these initial certs last, in practice they are disposed of within seconds.

That being said, the `DefaultRenewableCertTTL` is way too long, I'll set it to 1 hour.

r0mant commented on lib/services/local/users.go:
```diff
@@ -205,6 +205,51 @@ func (s *IdentityService) UpsertUser(user types.User) error {
+	if err := services.ValidateUser(new); err != nil {
+		return trace.Wrap(err)
+	}
+	newValue, err := services.MarshalUser(new.WithoutSecrets().(types.User))
```
Do a safe type assertion to avoid panics (just in case).

r0mant commented on lib/services/local/users.go:
```diff
@@ -205,6 +205,51 @@ func (s *IdentityService) UpsertUser(user types.User) error {
+		ID:      new.GetResourceID(),
+	}
+
+	existingValue, err := services.MarshalUser(existing.WithoutSecrets().(types.User))
```
Do a safe type assertion to avoid panics (just in case).

timothyb89 committed:
Another pass of review feedback

Ensure all requestable roles exist when creating a bot, adjust the
default renewable cert TTL down to 1 hour, and check types during
`CompareAndSwapUser()`

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot
